android/KMEA/app/src/main/java/com/keyman/engine/KMManager.java

```diff
@@ -631,32 +629,28 @@ public static RelativeLayout.LayoutParams getKeyboardLayoutParams() {
    return params;
   }
 
-  private static void initInAppKeyboard(Context appContext) {
-    if (InAppKeyboard == null) {
+  private static void initKeyboard(Context appContext, KeyboardType keyboardType) {
+    if (keyboardType == KeyboardType.KEYBOARD_TYPE_INAPP && InAppKeyboard == null) {
       InAppKeyboard = new KMKeyboard(appContext, KeyboardType.KEYBOARD_TYPE_INAPP);
       RelativeLayout.LayoutParams params = getKeyboardLayoutParams();
       InAppKeyboard.setLayoutParams(params);
       InAppKeyboard.setVerticalScrollBarEnabled(false);
       InAppKeyboard.setHorizontalScrollBarEnabled(false);
       InAppKeyboardWebViewClient = new KMKeyboardWebViewClient(appContext, KeyboardType.KEYBOARD_TYPE_INAPP);
       InAppKeyboard.setWebViewClient(InAppKeyboardWebViewClient);
-      InAppKeyboard.addJavascriptInterface(new KMInAppKeyboardJSHandler(appContext, InAppKeyboard), "jsInterface");
+      InAppKeyboard.addJavascriptInterface(new KMKeyboardJSHandler(appContext, InAppKeyboard), "jsInterface");
       InAppKeyboard.loadKeyboard();
 
       setEngineWebViewVersionStatus(appContext, InAppKeyboard);
-    }
-  }
-
-  private static void initSystemKeyboard(Context appContext) {
-    if (SystemKeyboard == null) {
+    } else if (keyboardType == KeyboardType.KEYBOARD_TYPE_SYSTEM && SystemKeyboard == null) {
       SystemKeyboard = new KMKeyboard(appContext, KeyboardType.KEYBOARD_TYPE_SYSTEM);
       RelativeLayout.LayoutParams params = getKeyboardLayoutParams();
       SystemKeyboard.setLayoutParams(params);
       SystemKeyboard.setVerticalScrollBarEnabled(false);
       SystemKeyboard.setHorizontalScrollBarEnabled(false);
       SystemKeyboardWebViewClient = new KMKeyboardWebViewClient(appContext, KeyboardType.KEYBOARD_TYPE_SYSTEM);
       SystemKeyboard.setWebViewClient(SystemKeyboardWebViewClient);
-      SystemKeyboard.addJavascriptInterface(new KMSystemKeyboardJSHandler(appContext, SystemKeyboard), "jsInterface");
+      SystemKeyboard.addJavascriptInterface(new KMKeyboardJSHandler(appContext, SystemKeyboard), "jsInterface");
       SystemKeyboard.loadKeyboard();
 
       setEngineWebViewVersionStatus(appContext, SystemKeyboard);
```

Comment:
This can be consolidated a bit further...

```suggestion
    Keyboard keyboard = null;
    KMKeyboardWebViewClient webViewClient = null;
    
    if(keyboardType == KeyboardType.KEYBOARD_TYPE_INAPP && InAppKeyboard == null) {
      InAppKeyboard = new KMKeyboard(appContext, keyboardType);
      InAppKeyboardWebViewClient = new KMKeyboardWebViewClient(appContext, keyboardType);
      keyboard = InAppKeyboard;
      webViewClient = InAppKeyboardWebViewClient;
    } else if(keyboardType == KeyboardType.KEYBOARD_TYPE_SYSTEM && SystemKeyboard == null) {
      SystemKeyboard = new KMKeyboard(appContext, keyboardType);    
      SystemKeyboardWebViewClient = new KMKeyboardWebViewClient(appContext, keyboardType);
      keyboard = SystemKeyboard;
      webViewClient = SystemKeyboardWebViewClient;
    }
    
    if(keyboard == null) {
      return;
    }
    
    RelativeLayout.LayoutParams params = getKeyboardLayoutParams();
    keyboard.setLayoutParams(params);
    keyboard.setVerticalScrollBarEnabled(false);
    keyboard.setHorizontalScrollBarEnabled(false);
    keyboard.setWebViewClient(webViewClient);
    keyboard.addJavascriptInterface(new KMKeyboardJSHandler(appContext, keyboard), "jsInterface");
    keyboard.loadKeyboard();

    setEngineWebViewVersionStatus(appContext, keyboard);
```

The only remaining reasons for the if-branching here:
- To verify that the keyboard we're initializing hasn't already been initialized.
- To assign our in-app- vs system- specific global variables their values.
    - If it weren't for that, we wouldn't need the if-else at all to initialize the Keyboard!

Reply:
I manually edited to incorporate your suggestion. (GitHub wouldn't let me accept it from the PR) 😕